source/source_integration_test.go

```diff
@@ -0,0 +1,436 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/brianvoe/gofakeit/v6"
+	"github.com/conduitio-labs/conduit-connector-hubspot/config"
+	"github.com/conduitio-labs/conduit-connector-hubspot/hubspot"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/matryer/is"
+)
+
+// testResource is a test resource that we use for integration tests.
+const testResource = "crm.contacts"
+
+var (
+	// testAccessToken will be used if a provided access token is empty,
+	// if both a provided access token and this value are empty an integration test will be skipped.
+	testAccessToken = os.Getenv("HUBSPOT_ACCESS_TOKEN")
+	// testHTTPClientTimeout is a HTTP timeout for test HTTP client.
+	testHTTPClientTimeout = 5 * time.Second
+
+	// consts for retries.
+	maxCheckRetries   = 5
+	checkRetryTimeout = time.Second * 2
+)
+
+func TestSource_Read_successSnapshot(t *testing.T) {
+	is := is.New(t)
+
+	// prepare a config, configure and open a new source
+	config := prepareConfig(t, "")
+
+	source := NewSource()
+
+	ctx := context.Background()
+
+	err := source.Configure(ctx, config)
+	is.NoErr(err)
+
+	// create a test hubspot client
+	hubspotClient := hubspot.NewClient(testAccessToken, &http.Client{
+		Timeout: testHTTPClientTimeout,
+	})
+
+	// create a test contact with a random properties
+	testContact, err := createTestContact(ctx, hubspotClient)
+	is.NoErr(err)
+
+	// give HubSpot some time to process the HTTP request we sent
+	// and create the contact
+	ok, err := waitTestContacts(ctx, hubspotClient, 1)
+	is.NoErr(err)
+	is.True(ok)
+
+	// open the source
+	err = source.Open(ctx, nil)
+	is.NoErr(err)
+
+	// read a record
+	record, err := source.Read(ctx)
+	is.NoErr(err)
+
+	is.Equal(record.Operation, sdk.OperationSnapshot)
+
+	testContactID := compareTestContactWithRecord(t, is, testContact, record)
+	err = hubspotClient.Delete(ctx, testResource, testContactID)
+	is.NoErr(err)
+
+	err = source.Teardown(ctx)
```

Comment:
Shouldn't we close the context above and call the `Teardown` method with the new one? 🤔

Reply:
Yeah, this definitely will be better. 👍